sql: integrate jsonpath parser with jsonpath type

normanchenn opened the pull request:
Previously, the `jsonpath` data type and `jsonpath` parsers were created. This PR integrates the parser with the data type.

Epic: None
Release note (sql change): Adding parser for jsonpath type. Accepts setting mode (`strict/lax`), key accessors (`.name`), and array wildcards (`[*]`).

normanchenn committed:
sql: accept uppercase characters in keys during jsonpath scanning

Previously, during jsonpath scanning, keys with uppercase characters
would be converted to lowercase. As per postgres, uppercase characters
are now accepted for key accessors.

Epic: None
Release note: None

yuzefovich reviewed (commented):
Nice! A few questions from me.

> Reviewed 4 of 4 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-:-OK7NV7p7HRiXgEyIS9Q:b-c8c03p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn)

___
*[`pkg/sql/scanner/scan.go` line 609 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-OK7MP0V5JIuNhVrqsM6:-OK7MP0V5JIuNhVrqsM7:bia65x5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1286a18e2482b9df601febaf668ec293effee1/pkg/sql/scanner/scan.go#L609)):*
> ```Go
> }
> 
> func (s *Scanner) normalizeIdent(lval ScanSymType, isIdentMiddle func(int) bool) {
> ```

nit: rather than duplicating most of the logic from `lowerCaseAndNormalizeIdent`, let's rename existing function to `normalizeIdent` and add an extra boolean argument that controls whether all ASCII characters are converted to lower case.
___
*[`pkg/sql/logictest/testdata/logic_test/jsonpath` line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-OK7NJLTCcVmTYMaAQ_p:-OK7NJLTCcVmTYMaAQ_q:bxvq00k) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc1dcfbcfeb0fb86e53f013befb4840db84d289b/pkg/sql/logictest/testdata/logic_test/jsonpath#L69)):*
> ```
> 
> statement error pq: could not parse "\$\.1a\[\*\]" as type jsonpath: at or near "1": syntax error
> SELECT '$.1a[*]'::JSONPATH
> ```

nit: this is parsable in PG, do we not support parsing this yet? Consider adding a TODO to indicate that.
___
*[`pkg/sql/randgen/datum.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-OK7LgrF8j1LuZDHpW-v:-OK7LgrF8j1LuZDHpW-w:bqeylo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1286a18e2482b9df601febaf668ec293effee1/pkg/sql/randgen/datum.go#L579)):*
> ```Go
> }
> 
> const charSet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
> ```

nit: why removing digits?


<!-- Sent from Reviewable.io -->

normanchenn reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142010)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @normanchenn and @yuzefovich)

___
*[`pkg/sql/randgen/datum.go` line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-OK7LgrF8j1LuZDHpW-v:-OK7O3vbE9HlIQA_BTgf:bpsv2pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f1286a18e2482b9df601febaf668ec293effee1/pkg/sql/randgen/datum.go#L579)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why removing digits?
</blockquote></details>

related to not being able to parse `$.1e`, the randomized function could generate a jsonpath expression with a digit as the first character of the key accessor so I removed the digits from the charSet for simplicity
___
*[`pkg/sql/logictest/testdata/logic_test/jsonpath` line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-OK7NJLTCcVmTYMaAQ_p:-OK7O9xl0bjNjEetBTR6:b-mjt4dh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc1dcfbcfeb0fb86e53f013befb4840db84d289b/pkg/sql/logictest/testdata/logic_test/jsonpath#L69)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this is parsable in PG, do we not support parsing this yet? Consider adding a TODO to indicate that.
</blockquote></details>

Yeah, we don't support it yet - I have a TODO in `pkg/sql/scanner/jsonpath_scan.go#L26` about this, basically because postgres will not parse `$.1e` due to the `e` being interpreted as part of a number.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-:-OK8E-gVETDaS6hoF5PS:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @normanchenn)

___
*[`pkg/sql/scanner/jsonpath_scan.go` line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-OK8Dkd577eJC6MjpNCn:-OK8Dkd577eJC6MjpNCo:bv8g9uy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b32977ca216d2664190364c57de8d9a64d3e024/pkg/sql/scanner/jsonpath_scan.go#L44)):*
> ```Go
> // scanIdent is similar to Scanner.scanIdent, but uses Jsonpath tokens.
> func (s *JSONPathScanner) scanIdent(lval ScanSymType) {
> 	s.normalizeIdent(lval, isIdentMiddle, false)
> ```

nit: let's add inline comment `/* toLower */` to all callsites.
___
*[`pkg/sql/scanner/scan.go` line 609 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/142010#-OK8DXkSCMrhB1SZ4ZuQ:-OK8DXkSCMrhB1SZ4ZuR:b-zdw24o) ([raw file](https://github.com/cockroachdb/cockroach/blob/54899c1d0eaf18a2dc2736406525c510d2a9b040/pkg/sql/scanner/scan.go#L609)):*
> ```Go
> }
> 
> func (s *Scanner) normalizeIdent(lval ScanSymType, isIdentMiddle func(int) bool) {
> ```

nit: let's move the change to `normalizeIdent` into the first commit. Also let's add a quick comment what `toLower` argument controls.


<!-- Sent from Reviewable.io -->

normanchenn committed:
sql: integrate jsonpath parser with jsonpath type

Previously, the `jsonpath` data type and `jsonpath` parsers were
created. This PR integrates the parser with the data type.

Epic: None
Release note (sql change): Adding parser for jsonpath type. Accepts
setting mode (`strict/lax`), key accessors (`.name`), and array wildcards
(`[*]`).

normanchenn commented:
TFTR!

bors r+
